### PR TITLE
2.6 mce hypershift 0825

### PR DESCRIPTION
**Pull request by rokej**
https://github.com/stolostron/backlog/issues/24993

https://github.com/stolostron/backlog/issues/25239

I also added doc on a couple of extra features we added in MCE 2.1.

### CODE REVIEW

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 1]
```diff
@@ -1,14 +1,14 @@
 [#hosted-control-planes-configure]
 = Configuring hosted control planes
 
-Configuring hosted control planes requires a hosting service cluster and a hosted cluster. By deploying the HyperShift operator on an existing cluster, you can make that cluster into a hosting service cluster and start the creation of the hosted cluster. 
+Configuring hosted control planes requires a hosting service cluster. By deploying the HyperShift operator on an existing cluster, you can make that cluster into a hosting service cluster and start the creation of the hosted cluster.
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Are you sure you don't need a hosted cluster? In the HyperShift doc it says you do.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 2]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
 oc label secret hypershift-operator-oidc-provider-s3-credentials -n <hypershift-hosting-service-cluster> cluster.open-cluster-management.io/backup=""
 ----
 
+. If you plan to provision hosted clusters on the AWS platform with Private Link, create an AWS credential secret for the HyperShift operator, and name it `hypershift-operator-private-link-credentials`. It should reside in the managed cluster namespace (i.e., the namespace of the managed cluster that will be used as the hosting service cluster). If you used `local-cluster`, then create the secret in the `local-cluster` namespace. See steps 1-5 in https://hypershift-docs.netlify.app/how-to/aws/deploy-aws-private-clusters/[Deploy AWS private clusters].
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Replace i.e. with "that is" to adhere to IBM style.

Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: "It should reside in the managed cluster namespace"

We need to be more direct and avoid using "should". Is that namespace where it must be or is it optional? We can say it must reside in the managed cluster namespace if it has to be there.

Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: For the last sentence, change it to something like this:

If you are using `local-cluster`, create the secret in the `local-cluster` namespace. See  steps 1–5 in https://hypershift-docs.netlify.app/how-to/aws/deploy-aws-private-clusters/[Deploy AWS private clusters] for more details.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 3]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
 oc label secret hypershift-operator-oidc-provider-s3-credentials -n <hypershift-hosting-service-cluster> cluster.open-cluster-management.io/backup=""
 ----
 
+. If you plan to provision hosted clusters on the AWS platform with Private Link, create an AWS credential secret for the HyperShift operator, and name it `hypershift-operator-private-link-credentials`. It should reside in the managed cluster namespace (i.e., the namespace of the managed cluster that will be used as the hosting service cluster). If you used `local-cluster`, then create the secret in the `local-cluster` namespace. See steps 1-5 in https://hypershift-docs.netlify.app/how-to/aws/deploy-aws-private-clusters/[Deploy AWS private clusters]. 
++
+Instructions result in the following AWS resources:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Which instructions are you referring to? If the following is what happens after completing steps 1 through 5, I would say something like: After creating the secret, the following AWS resources appear:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 4]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
++
+* Create an IAM policy in AWS
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: I'll review the bullet list again once the previous comment has been addressed.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 5]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
++
+* Create an IAM user for the Hypershift operator to use
++
+* Attach the IAM policy to the IAM user
++
+* Create an IAM access key for the user, use the access key and secret in the next section
++
+The secret must contain 3 fields:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: The secret must contain the following 3 fields:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 6]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
++
+* `aws-access-key-id`: AWS credential access key id
++
+* `aws-secret-access-key`: AWS credential access key secret
++
+* `region`: Region for use with Private Link
++
+See https://hypershift-docs.netlify.app/getting-started/[Getting started] in the HyperShift documentation for more information about the secret. The following example shows sample `hypershift-operator-private-link-credentials` secret template:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: The following example shows the sample `hypershift-operator-private-link-credentials` secret template:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 7]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
+oc create secret generic hypershift-operator-private-link-credentials --from-literal=aws-access-key-id=<aws-access-key-id> --from-literal=aws-secret-access-key=<aws-secret-access-key> --from-literal=region=<region> -n <hypershift-hosting-service-cluster>
+----
++
+*Note:* Disaster recovery backup for the secret is not automatically enabled. Run the following command to add the label that enables the `hypershift-operator-private-link-credentials` secret to be backed up for disaster recovery:
++
+----
+oc label secret hypershift-operator-private-link-credentials -n <hypershift-hosting-service-cluster> cluster.open-cluster-management.io/backup=""
+----
++
+Set the following parameter in HypershiftDeployment, when creating a cluster:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Is that a file name? If so, use `HypershiftDeployment`

Remove the comma as well.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 8]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
+----
+spec:
+  hostedClusterSpec:
+    platform:
+      type: AWS
+      aws:
+        endpointAccess: Private
+----
++
+. If you plan to use service-level DNS for Control Plane Service on AWS platform, create an external DNS credential secret for the HyperShift operator, and name it hypershift-operator-external-dns-credentials. It should reside in the managed cluster namespace (i.e., the namespace of the managed cluster that will be used as the hosting service cluster). If you used local-cluster, then create the secret in the local-cluster namespace
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: If you plan to use service-level DNS for control plane services on the AWS platform, create an external DNS credential secret for the HyperShift operator and name it `hypershift-operator-external-dns-credentials`.

See previous comment as well, don't use "should" and replace the i.e. Change to present tense in last sentence.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc (hunk 2 above)
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Remove comma before "and"

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 9]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
++
+The secret must contain 3 fields:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: The secret must contain the following 3 fields:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 10]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
++
+* `domain-filter`: The service-level domain
++
+* `credentials`: (Optional, only when using aws keys) - For all external DNS types, a credential file is supported
++
+* `aws-access-key-id`: (Optional) - When using AWS DNS service, credential access key id
++
+* `aws-secret-access-key`: (Optional) - When using AWS DNS service, When using AWS DNS service, credential access key secret
++
+See https://hypershift-docs.netlify.app/how-to/external-dns/[HyperShift Project Documentation] for details. The following example shows sample `hypershift-operator-external-dns-credentials` secret template:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: See https://hypershift-docs.netlify.app/how-to/external-dns/[HyperShift Project Documentation] for more details. The following example shows the sample `hypershift-operator-external-dns-credentials` secret template:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 11]
```diff
@@ -71,6 +71,88 @@ oc create secret generic hypershift-operator-oidc-provider-s3-credentials --from
...
+oc create secret generic hypershift-operator-external-dns-credentials --from-literal=provider=aws --from-literal=domain-filter=service.my.domain.com --from-file=credentials=<credentials-file> -n <hypershift-hosting-service-cluster>
+----
++
+*Note:* Disaster recovery backup for the secret is not automatically enabled. Run the following command to add the label that enables the `hypershift-operator-external-dns-credentials` secret to be backed up for disaster recovery:
++
+----
+oc label secret hypershift-operator-external-dns-credentials -n <hypershift-hosting-service-cluster> cluster.open-cluster-management.io/backup=""
+----
++
+Set the following parameter in HypershiftDeployment, when creating a cluster:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: If HypershiftDeployment is a name for a file, use `HypershiftDeployment`.

Remove comma as well.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 12]
```diff
@@ -153,10 +235,70 @@ oc create secret generic <my-secret> -n <hypershift-deployment-namespace> --from
...
++
+You can refer to the https://github.com/stolostron/hypershift-deployment-controller/blob/main/api/v1alpha1/hypershiftdeployment_types.go[field definitions] of the API to ensure that they are correct.
+
+. Check the `HypershiftDeployment` status by running the following command:
++
+----
+oc get hypershiftdeployment -n default hypershift-demo -w
+----
+
+. After the hosted cluster is created, it is automatically imported to the hub. You can verify this by viewing the cluster list in the {product-title-short} console, or by running the following command:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: To verify that the hosted cluster is imported, view the cluster list in the {product-title-short} console, or by running the following command:

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 13]
```diff
@@ -153,10 +235,70 @@ oc create secret generic <my-secret> -n <hypershift-deployment-namespace> --from
...
+. After the hosted cluster is created, it is automatically imported to the hub. You can verify this by viewing the cluster list in the {product-title-short} console, or by running the following command: 
++
+----
+oc get managedcluster <hypershiftDeployment.Spec.infraID>
+----
+
+[#hosting-service-cluster-access]
+== Accessing a hosting service cluster
+
+You can now access your cluster. The access secrets are stored in the `hypershift-hosting-service-cluster` namespace. This namespace is the same as the name of the hosting service cluster. Learn about the following formats secret name formats:
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Not sure if the first sentence is necessary. I would probably remove it.

-> See the following format secret name formats to learn more.

^I don't really like "format secret name formats", can we write around that somehow or phrase it differently?

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc [hunk 14]
```diff
@@ -153,10 +235,70 @@ oc create secret generic <my-secret> -n <hypershift-deployment-namespace> --from
...
+
+- `kubeconfig` secret: `<hypershiftDeployment.Spec.hostingNamespace>-<hypershiftDeployment.Name>-admin-kubeconfig` (clusters-hypershift-demo-admin-kubeconfig)
+- `kubeadmin` password secret: `<hypershiftDeployment.Spec.hostingNamespace>-<hypershiftDeployment.Name>-kubeadmin-password` (clusters-hypershift-demo-kubeadmin-password)
+
+[#customizing-hosted-cluster-specifications]
+== Customizing hosted cluster specifications in HypershiftDeployment
+
+In a `HypershiftDeployment` custom resource, you can change `hostedcluster` and `nodepool` specifications. For example, you can change the OCP release image of the hosted cluster control plane and/or the nodepool, the management spec of the nodepool or the number of nodes in the nodepool.
```
Comment (oafischer) [inline multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Add comma before the last "or".

**Review by oafischer — CHANGES_REQUESTED** [review]
Hi @rokej, thanks for submitting this PR! I've left a few comments for you to work through. Let me know if you have any questions. You can always reach me on Slack as well.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc (hunk 1 above)
Comment (rokej) [reply multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Hosted cluster is hosted control planes but I guess it also makes sense to require a hosted cluster (custom resource) to configure hosted control planes. I added it back.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc (hunk 2 above)
Comment (rokej) [reply multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: must

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc (hunk 4 above)
Comment (rokej) [reply multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: Not needed.

File: multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc (hunk 13 above)
Comment (rokej) [reply multicluster_engine/hosted_control_planes/hosted_control_planes_configure.adoc]: How about..


The access secret name formats are:

**Comment by rokej** [comment]
@oafischer Thank you so much for your review comments. I addressed all of them. Please review and and let me know. Thank you.

**Review by oafischer — APPROVED** [review]
@rokej Thanks for working through the comments! I fixed a couple of typos. Otherwise it looks good to me. Approving.

**Pull request merged**